firmware/ventilator-controller-stm32/Core/Src/Pufferfish/Driver/Serial/Nonin/PacketReceiver.cpp

```diff
@@ -33,214 +33,93 @@ namespace Pufferfish::Driver::Serial::Nonin {
-   * bit0
-   */
-  sensor_measurements.e_spo2 = get_6bit_data(packet_data[e_spo2_index][byte3]);
-
-  /**
-   * E-spo2-D: 8-beat average displayed values in display mode
-   * sensorData 17th frame of packet contains spo2-D-Bit6 to E-spo2-D-Bit0 in
-   * bit6 to bit0
-   */
-  sensor_measurements.e_spo2_d = get_6bit_data(packet_data[e_spo2_d_index][byte3]);
-
-  /**
-   * HR-D: 8-beat average values in standard mode.
-   * HR-D MSB : sensorData 20th frame of packet contains HR-Bit8 and HR-Bit7 in
-   * bit1 and bit0 HR-D LSB : sensorData 21st frame of packet contains HR-Bit6
-   * t0 HR-Bit0 in bit6 to bit0
-   */
-  sensor_measurements.heart_rate_d = get_9bit_data(
-      packet_data[heart_rate_d_index][byte3], packet_data[heart_rate_d_index + 1][byte3]);
-
-  /**
-   * E-HR-D: 8-beat average displayed values in display mode
-   * E-HR-D MSB : sensorData 22nd frame of packet contains HR-Bit8 and HR-Bit7
-   * in bit1 and bit0 E-HR-D LSB : sensorData 23rd frame of packet contains
-   * HR-Bit6 t0 HR-Bit0 in bit6 to bit0
-   */
-  sensor_measurements.e_heart_rate_d = get_9bit_data(
-      packet_data[e_heart_rate_d_index][byte3], packet_data[e_heart_rate_d_index + 1][byte3]);
-
-  /* Update the PLETH and Status byte errors into the packet measurements */
-  for (frame_index = 0; frame_index < packet_size; frame_index++) {
+
+  // SpO2 data
+  sensor_measurements.spo2 = get_spo2_data(packet_data[spo2_index][byte3]);
+  sensor_measurements.spo2_d = get_spo2_data(packet_data[spo2_d_index][byte3]);
+  sensor_measurements.spo2_fast = get_spo2_data(packet_data[spo2_fast_index][byte3]);
+  sensor_measurements.spo2_b_b = get_spo2_data(packet_data[spo2_b_b_index][byte3]);
+  sensor_measurements.e_spo2 = get_spo2_data(packet_data[e_spo2_index][byte3]);
+  sensor_measurements.e_spo2_d = get_spo2_data(packet_data[e_spo2_d_index][byte3]);
+  // Firmware revision
+  sensor_measurements.firmware_revision = packet_data[firmware_revision_index][byte3];
+  // HR data
+  sensor_measurements.hr =
+      get_hr_data(packet_data[hr_index][byte3], packet_data[hr_index + 1][byte3]);
+  sensor_measurements.e_hr =
+      get_hr_data(packet_data[e_hr_index][byte3], packet_data[e_hr_index + 1][byte3]);
+  sensor_measurements.hr_d =
+      get_hr_data(packet_data[hr_d_index][byte3], packet_data[hr_d_index + 1][byte3]);
+  sensor_measurements.e_hr_d =
+      get_hr_data(packet_data[e_hr_d_index][byte3], packet_data[e_hr_d_index + 1][byte3]);
+
+  for (size_t frame_index = 0; frame_index < packet_size; ++frame_index) {
     sensor_measurements.packet_pleth[frame_index] = packet_data[frame_index][byte2];
     read_status_byte(sensor_measurements, frame_index, packet_data[frame_index][byte1]);
   }
 }
 
-PacketReceiver::PacketInputStatus PacketReceiver::input(const Frame &frame) {
-  /* Check the frame received is first frame in the packet SYNC bit is 1 */
+PacketInputStatus PacketReceiver::input(const Frame &frame) {
   if ((frame[1] & 0x01U) == 0x01) {
```

Comment:
Since this `if` statement is changing a lot of internal state (which is where our codebase is most error-prone), we should use descriptive naming and code comments to help communicate what's going on.

What does the expression being tested by the `if` statement actually mean? We can answer this question in the code by having a descriptive function name to replace `(frame[1] & 0x01U) == 0x01`.

What does this `if` statement try to achieve when the condition is met? We can answer this question in the code by adding a code comment which summarizes everything inside, as a replacement for the line-by-line code comments which take too much work to try to compile into a summary-level understanding.